### ISSUE TITLE
System.Numerics.BigInteger causes StackOverflowException

### ISSUE THREAD

**Issue by ArshBunny** [dotnet/corefx#13568]
![Screenshot](http://archive.org/download/Fsharp-StackOverflowException/fsi.png)

The F# code first defines a custom power function, then tetration function based on it and then two operators so I don't have to type 'I' next to every number I want to work with and hide the superfluous BigInteger members from the console.

It worked fine for 3rd tetration of 4 [4 ^ (4 ^ 4) = 4 ^ 256] and returned over 150 digits, but something went wrong on calculating the **5th tetration of 2**, which **is supposed to be almost 20,000 digits**, btw; the program crashes before bigint value gets close to what float can handle (~310 digits), let alone 20,000 digits. What am I missing?
```
let rec exponentiation (x:bigint) (y:bigint) : bigint =
    match y with
    | y when y.IsZero -> 1I
    | y when y.IsOne -> x
    | y -> x * exponentiation x (y-1I)
let rec tetration (x:bigint) (y:bigint) : bigint =  
    match y with
    | y when y.IsZero -> 1I
    | y when y.IsOne -> x
    | y -> exponentiation x (tetration x (y-1I))
let (^) (x:int) (y:int) = System.Console.WriteLine (exponentiation (bigint x) (bigint y))
let (^^) (x:int) (y:int) = System.Console.WriteLine (tetration (bigint x) (bigint y))
```
I've been dabbling with F# lately, it's the only programming language I am familiar with. I know an unhelpful smattering of JavaScript, and remember a couple of keywords from C# that I tried years ago for a few days on Unity (yeah, absolutely no .NET knowledge)
* Windows 10 for Education x64
* .NET Framework 4.5.1
* Visual F# 4.0
* Visual Studio Code 1.7 with Ionide suite of extensions

**Comment by Clockwork-Muse** [comment]
Your expoentiation function isn't tail-call recursive.  Are you sure its not just regular stack overflow?

**Comment by Clockwork-Muse** [comment]
TL;DR - the multiplication operation isn't considered for tail-call optimizations.  I'm not sure what to do on that end, or if anything should be done, given it's a heavy operation.

<br />

Yeah, looks like there's no problem in `BigInteger`, it handles that multiplication just fine.  You're not even going to be touching the recursive algorithms there, it's going to be a simple loop.  The problem is your F# code, because you're not using a form it can change into a tail-call.

Okay, so why is it overflowing?  First, we can prove that it's not doing tail recursion by looking at the disassembled IL, which contains something like this:
```
....some lines removed/shortened....
IL_003d:  call valuetype [System.Numerics].BigInteger [System.Numerics].BigInteger::get_One()
IL_0042:  call valuetype [System.Numerics].BigInteger [System.Numerics].BigInteger::op_Subtraction(valuetype [System.Numerics].BigInteger, valuetype [System.Numerics].BigInteger)
IL_0047:  call valuetype [System.Numerics].BigInteger Program::exponentiation(valuetype [System.Numerics.BigInteger, valuetype [System.Numerics].BigInteger)
IL_004c:  call valuetype [System.Numerics].BigInteger [System.Numerics].BigInteger::op_Multiply(valuetype [System.Numerics].BigInteger, valuetype [System.Numerics].BigInteger)
IL_0051:  ret
```
...note that the recursive call to `exponentiation` happens as a separate step to the multiplication and return.

Because of this, for the final `exponentiation` call from `tetration`, we know the stack is going to be 65,536 frames deep.  At each frame we are storing (at minimum): `x` and the result of the subtraction, `y - 1I`.  Each instance of both of these require 4 bytes for the sign part (+ data, since the values are small enough), and 4 (or 8) bytes for the null reference to the data array.  So, 2 * (4 + 4) * 65,536 = 1,048,576, or 1MB, which is the default size of the stack.  That it overflows is hardly surprising.

So what can you do about it?  Two quick things to do.
 1. Use the appropriate method, if possible.  There _is_ a `Pow` method, after all.
 2. Use an accumulator, so the compiler can do a better job of optimization.

You probably want to start with something that looks closer to this:
```
let rec exponentiation2(x:bigint) (y:bigint) : bigint =
    let maxVal = bigint System.Int32.MaxValue // pulled out or it _may_ new it up each time
    let rec innerExponentiation (x:bigint) (y:bigint) (acc:bigint) :bigint =
        match y with
        | y when y <= maxVal -> acc * bigint.Pow(x, int y)
        | y -> innerExponentiation x (y - 1I) (acc * x)
    innerExponentiation x (y - 1I) x
```
...for most common cases, it's just going to shortcut and call `Pow`, which is almost certainly faster than performing the multiplication multiple times.  If you somehow have a larger number (...where you'd probably hit OOM due to not getting a contiguous block for the array), using an accumulator makes the optimizer turn the recursive form into a loop.

**Comment by stephentoub** [comment]
@Clockwork-Muse, thanks for the investigation.

@ArshBunny, it sounds like the issue isn't actually in BigInteger.  Can this be closed?

**Comment by ArshBunny** [comment]
The issue had to with my code and the stack, and not with the limitation of System.BigInteger as I initially thought. This is the code I ended up with (using accumulator pattern for the exponentiation function):
```
let rec exponentiation (x:bigint) (y:bigint) (acc:bigint) : bigint =
    match y with
    | y when y.IsZero -> acc
    | y when y.IsOne -> x * acc
    | y -> exponentiation x (y - 1I) (x * acc)
    | _ -> failwith "Negative powers not supported."
let rec tetration (x:bigint) (y:bigint) : bigint =  
    match y with
    | y when y.IsZero -> 1I
    | y when y.IsOne -> x
    | y -> exponentiation x (tetration x (y - 1I)) (bigint 1)
    | _ -> failwith "Negative height-parameters not supported."
let (^) (x:int) (y:int) = System.Console.WriteLine (exponentiation (bigint x) (bigint y) (bigint 1))
let (^^) (x:int) (y:int) = System.Console.WriteLine (tetration (bigint x) (bigint y))
```
I shyed away from `Pow()` because I only knew it to work with float values and adding ".0" after every number became tedious, and also [the number I was working to obtain](https://archive.org/download/twocaretcaretfive/2%5e%5e5.png) was a little longer than 350 digits.

**Comment by Clockwork-Muse** [comment]
@ArshBunny - I meant, use `BigInteger.Pow`, which doesn't deal with floats.
Also, you don't have to subtract and multiply by a single operator each time; a logn solution is possible (an exercise left to the reader).
